### PR TITLE
Feature/network

### PR DESCRIPTION
**Pull request by xiangbingj**
New features 

- [x] I have thoroughly tested my contribution.
- [x] The code I submitted has no copyright issues.

\<Description of and rational behind this PR\>Adding network drivers.

### CODE REVIEW

File: lib/freertos/include/kernel/driver.hpp [hunk 1]
```diff
@@ -384,6 +384,46 @@ class filesystem_file : public virtual object_access
...
+{
+public:
+    virtual object_accessor<network_socket> accept(socket_address_t *remote_address) = 0;
+    virtual void bind(const socket_address_t &address) = 0;
+    virtual void connect(const socket_address_t &address) = 0;
+    virtual void listen(uint32_t backlog) = 0;
+    virtual void shutdown(socket_shutdown_t how) = 0;
+    virtual size_t send(gsl::span<const uint8_t> buffer, uint8_t flags) = 0;
+    virtual size_t receive(gsl::span<uint8_t> buffer, uint8_t flags) = 0;
+    virtual size_t send_to(gsl::span<const uint8_t> buffer, uint8_t flags, const socket_address_t *to) = 0;
```
Comment (sunnycase) [inline lib/freertos/include/kernel/driver.hpp]: Use reference instead of pointer.

File: lib/freertos/include/network.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+int network_socket_send_to(handle_t socket_handle, const uint8_t *data, size_t len, uint8_t flags, const socket_address_t *to);
+
+int network_socket_receive_from(handle_t socket_handle, uint8_t *data, size_t len, uint8_t flags, socket_address_t *from);
+
+hostent_t *network_socket_gethostbyname(const char *name);
+
+int network_socket_addr_parse(const char *ip_addr, int port, uint8_t *socket_addr);
+
+int network_socket_addr_to_string(uint8_t *socket_addr, char *ip_addr, int *port);
```
Comment (sunnycase) [inline lib/freertos/include/network.h]: uint8_t or char?

File: lib/freertos/include/osdefs.h [hunk 3]
```diff
@@ -246,6 +246,74 @@ typedef struct _find_file_data
...
+    uint8_t **h_aliases;
+    /* Address type. */ 
+    uint32_t h_addrtype;
+    /* The length, in bytes, of the address. */
+    uint32_t h_length;
+    /* A pointer to an array of pointers to network addresses (in
+    network byte order) for the host, terminated by a null pointer. */
+    uint8_t **h_addr_list; 
+    /* for backward compatibility */
+#define h_addr h_addr_list[0]
```
Comment (sunnycase) [inline lib/freertos/include/osdefs.h]: h_addr is not used?

File: lib/freertos/include/network.h (hunk 2 above)
Comment (xiangbingj) [reply lib/freertos/include/network.h]: string to _socket_address.data[].should be uint8_t

**Pull request merged**